File: epi_judge_java_solutions/epi/ConvertBase.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class ConvertBase {
  @EpiTest(testDataFile = "convert_base.tsv")

  public static String convertBase(String numAsString, int b1, int b2) {

    boolean isNegative = numAsString.startsWith("-");
    int numAsInt =
        numAsString.substring(isNegative ? 1 : 0)
            .chars()
            .reduce(0,
                    (x, c)
                        -> x * b1 +
                               (Character.isDigit(c) ? c - '0' : c - 'A' + 10));
    return (isNegative ? "-" : "") +
        (numAsInt == 0 ? "0" : constructFromBase(numAsInt, b2));
  }

  private static String constructFromBase(int numAsInt, int base) {
    return numAsInt == 0
        ? ""
        : constructFromBase(numAsInt / base, base) +
              (char)(numAsInt % base >= 10 ? 'A' + numAsInt % base - 10
                                           : '0' + numAsInt % base);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "ConvertBase.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python_solutions/convert_base.py
import functools
import string

from test_framework import generic_test


def convert_base(num_as_string, b1, b2):
    def construct_from_base(num_as_int, base):
        return ('' if num_as_int == 0 else
                construct_from_base(num_as_int // base, base) +
                string.hexdigits[num_as_int % base].upper())

    is_negative = num_as_string[0] == '-'
    num_as_int = functools.reduce(
        lambda x, c: x * b1 + string.hexdigits.index(c.lower()),
        num_as_string[is_negative:], 0)
    return ('-' if is_negative else '') + ('0' if num_as_int == 0 else
                                           construct_from_base(num_as_int, b2))


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("convert_base.py", "convert_base.tsv",
                                       convert_base))
